File: New_NativeUI/Items/UIMenuListItem.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Threading.Tasks;
using CitizenFX.Core.UI;

namespace ScaleformUI
{
    public class UIMenuListItem : UIMenuItem, IListItem
    {
        protected internal int _index;
        protected internal List<dynamic> _items;


        /// <summary>
        /// Triggered when the list is changed.
        /// </summary>
        public event ItemListEvent OnListChanged;

        /// <summary>		
        /// Triggered when a list item is selected.		
        /// </summary>		
        public event ItemListEvent OnListSelected;

        /// <summary>
        /// Returns the current selected index.
        /// </summary>
        public int Index
        {
			get { return _index % Items.Count; }
			set { _index = 100000000 - (100000000 % Items.Count) + value; }
		}

		/// <summary>
		/// Returns the current selected index.
		/// </summary>
		public List<object> Items
		{
			get => _items;
			set
			{
				Index = 0;
				_items = value;
			}
		}

		/// <summary>
		/// List item, with left/right arrows.
		/// </summary>
		/// <param name="text">Item label.</param>
		/// <param name="items">List that contains your items.</param>
		/// <param name="index">Index in the list. If unsure user 0.</param>
		public UIMenuListItem(string text, List<dynamic> items, int index) : this(text, items, index, "")
        {
        }

        /// <summary>
        /// List item, with left/right arrows.
        /// </summary>
        /// <param name="text">Item label.</param>
        /// <param name="items">List that contains your items.</param>
        /// <param name="index">Index in the list. If unsure user 0.</param>
        /// <param name="description">Description for this item.</param>
        public UIMenuListItem(string text, List<dynamic> items, int index, string description) : this(text, items, index, description, HudColor.HUD_COLOUR_PAUSE_BG, HudColor.HUD_COLOUR_WHITE)
        {
        }

        public UIMenuListItem(string text, List<dynamic> items, int index, string description, HudColor mainColor, HudColor higlightColor) : this(text, items, index, description, mainColor, higlightColor, HudColor.HUD_COLOUR_WHITE, HudColor.HUD_COLOUR_BLACK)
        {
        }
        
        public UIMenuListItem(string text, List<dynamic> items, int index, string description, HudColor mainColor, HudColor higlightColor, HudColor textColor, HudColor highlightTextColor) : base(text, description, mainColor, higlightColor, textColor, highlightTextColor)
		{
			const int y = 0;
			_items = items;
			Index = index;
			_itemId = 1;
		}


		/// <summary>
		/// Find an item in the list and return it's index.
		/// </summary>
		/// <param name="item">Item to search for.</param>
		/// <returns>Item index.</returns>
		[Obsolete("Use UIMenuListItem.Items.FindIndex(p => ReferenceEquals(p, item)) instead.")]
		public virtual int ItemToIndex(dynamic item)
        {
            return _items.FindIndex(p => ReferenceEquals(p, item));
        }

		/// <summary>
		/// Find an item by it's index and return the item.
		/// </summary>
		/// <param name="index">Item's index.</param>
		/// <returns>Item</returns>
		[Obsolete("Use UIMenuListItem.Items[Index] instead.")]
		public virtual dynamic IndexToItem(int index)
        {
            return _items[index];
        }

        /// <summary>
        /// Add a Panel to the UIMenuListItem
        /// </summary>
        /// <param name="panel"></param>
        public override void AddPanel(UIMenuPanel panel)
        {
            Panels.Add(panel);
            panel.SetParentItem(this);
        }

        internal virtual void ListChangedTrigger(int newindex)
        {
            OnListChanged?.Invoke(this, newindex);
        }

        internal virtual void ListSelectedTrigger(int newindex)
        {
            OnListSelected?.Invoke(this, newindex);
        }

        public void ChangeList(List<dynamic> list, int index)
        {
            _items.Clear();
            _items = list;
            ScaleformUI._ui.CallFunction("UPDATE_LISTITEM_LIST", Parent.MenuItems.IndexOf(this), string.Join(",", _items), index);
        }

        public override void SetRightBadge(BadgeIcon badge)
        {
            throw new Exception("UIMenuListItem cannot have a right badge.");
        }

        public override void SetRightLabel(string text)
        {
            throw new Exception("UIMenuListItem cannot have a right label.");
        }

		[Obsolete("Use UIMenuListItem.Items[Index].ToString() instead.")]
		public string CurrentItem()
        {
            return _items[Index].ToString();
        }
    }
}

File: ScaleformUI_V8/src/menus/UIMenu/items/uimenulistitem.ts
import { UIMenuItem } from "./uimenuitem";
import { ListItemChangeCallbackBuilder, UIMenuListItemChangeCallback } from '../emitters/emitters';
import { SColor } from "elements/scolor";
import { BadgeStyle } from "elements/badge";

export class UIMenuListItem extends UIMenuItem {
    private _index: number = 0;
    private _items: any[] = [];

    private _listChangedEmitter = new ListItemChangeCallbackBuilder();
    private _listSelectedEmitter = new ListItemChangeCallbackBuilder();

    public set Index(value: number) {
        if (value < 0)
            this._index = 0;
        else if (value >= this._items.length)
            this._index = this.Items.length - 1;
        else
            this._index = value;
        if (this.Parent !== null && this.Parent.Visible && this.Parent.Pagination.IsItemVisible(this.Parent.Items.indexOf(this))) {
            ScaleformUI.Scaleforms._ui.callFunction("SET_ITEM_VALUE", this.Parent.Pagination.GetScaleformIndex(this.Parent.Items.indexOf(this)), this._index);
        }
    }
    public get Index():number{
        return this._index % this.Items.length;
    }

    public set Items(value:any[]){
        this.Index = 0;
        this._items = value;
    }
    public get Items():any[]{
        return this._items;
    }

    constructor(text:string , items:any[] , index:number , description:string , mainColor:SColor , higlightColor:SColor , textColor:SColor , highlightTextColor: SColor){
        super(text, description, mainColor, higlightColor, textColor, highlightTextColor);
        this._itemId = 1;
        this._items = items;
        this.Index = index;
    }

    public ChangeList(list:any[], index:number){
        this._items.length = 0;
        this._items = list;
        this._index = index;
        if (this.Parent !== null && this.Parent.Visible && this.Parent.Pagination.IsItemVisible(this.Parent.Items.indexOf(this))) {
            ScaleformUI.Scaleforms._ui.callFunction("UPDATE_LISTITEM_LIST", this.Parent.Pagination.GetScaleformIndex(this.Parent.Items.indexOf(this)), this._items.join(","), index);
        }
    }

    public onListChanged(delegate: UIMenuListItemChangeCallback) {
        this._listChangedEmitter.add(delegate)
    }
    public onHighlighted(delegate: UIMenuListItemChangeCallback) {
        this._listSelectedEmitter.add(delegate)
    }

    listChangedEmit() {
        this._listChangedEmitter.toDelegate()(this, this.Index);
    }
    listSelectedEmit() {
        this._listSelectedEmitter.toDelegate()(this, this.Index);
    }

    public SetRightBadge(badge: BadgeStyle) {
        throw new Error("UIMenuListItem cannot have a right badge.");
    }
    
    public SetRightLabel(text: string) {
        throw new Error("UIMenuListItem cannot have a right label.");
    }

}